ast/builtins.go

```diff
@@ -1706,6 +1707,27 @@ var JSONMarshal = &Builtin{
 	Categories: encoding,
 }
 
+var JSONMarshalIndent = &Builtin{
+	Name: "json.marshal_indent",
+	Description: "Serializes the input term to indented JSON, including line breaks, with optional prefix. " +
+		"One or more copies of the indent string are emitted before each child JSON element, depending on indent level.",
+	Decl: types.NewFunction(
+		types.Args(
+			types.Named("x", types.A).Description("the term to serialize"),
+			types.Named("indent", types.NewAny(
```

Comment:
Instead of having two arguments, `indent` and `prefix`, might I instead suggest taking an `options` object as arg which contains these two string parameters (similar to [providers.aws.sign_req()](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-providersaws-providersawssign_req) or [http.send()](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-http-httpsend), kinda). 

This way, we gain a couple of benefits:

* `prefix` will be truly optional, and users don't need to provide a `null` value for it, which will likely be the most common usage.
* we're future-proofing this built-in, and can add new marshaling options (e.g string escaping) in the future without needing to bloat the list of built-ins.
* users (primarily reviewers) don't need to keep track of the argument order, which is likely just gonna force them to revisit the docs a bunch.

If we make the above change, then this new built-in wouldn't be confined to indentation, so a rename would be necessary. Naming is hard, so this is where I regret making this suggestion 😅. Maybe `json.marshal_with_options` or simply `json.marshal2` 🤔 .

Reply:
☝️ I like that. _Defaults_ can then be applied if the key `prefix` or `indent` isn't present in the `opts` object.

Reply:
I quite like `json.marshal_with_options` myself. Similar to this [recently merged one](https://github.com/open-policy-agent/opa/pull/6643).

I think that we do want an error for someone passing an unknown key, as most likely it's a typo or something from an LLM etc. I think that returning a simple error from the built-in like `unexpected marshal opt "indentt"` or similar would be a good course of action. Similar to: https://github.com/open-policy-agent/opa/blob/630b74683a01a398f899bcad00c3095fb4482f8b/topdown/crypto.go#L265


Re documentation, I think it'd be much preferred to copy what is used for [providers.aws.sign_req](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-providersaws-providersawssign_req) rather than `http.send`. Http send is a special case as it has just so many different options. I feel like the docs table for these is in need of some work, but for now if we follow `providers.aws.sign_req` that makes sense to me.

Reply:
If the typing is done correctly here where the built-in is registered, the built-in implementation shouldn't need to check any other attributes than those it's explicitly interested in. See e.g. the [json.patch builtin](https://github.com/open-policy-agent/opa/blob/d46bc9d2be7d9cc0e3a7016b01f8e263012dc736/ast/builtins.go#L1543).

Reply:
I can rename to `marshal_with_options`, sure.



> Re documentation, I think it'd be much preferred to copy what is used for [providers.aws.sign_req](https://www.openpolicyagent.org/docs/latest/policy-reference/#builtin-providersaws-providersawssign_req)

@charlieegan3 I think I might've miscommunicated what I was looking for. Looking at the example for AWS, I'm not sure what the process is for adding anything other than the AST-level docs for the builtin proper:
![image](https://github.com/open-policy-agent/opa/assets/72675818/16a97a61-2c3e-4814-afe0-ee6727c8c640)

Should I edit `docs\content\policy_reference.md` directly, or is there some other area where this documentation is built from?



> If the typing is done correctly here where the built-in is registered, the built-in implementation shouldn't need to check any other attributes than those it's explicitly interested in. See e.g. the [json.patch builtin](https://github.com/open-policy-agent/opa/blob/d46bc9d2be7d9cc0e3a7016b01f8e263012dc736/ast/builtins.go#L1543).

@johanfylling Static properties precludes setting defaults, though - right? (i.e. users would have to provide those properties on every object they pass to `marshal_with_options`)

Reply:
I... _think_ all of the requisite changes (minus doc updates) have been made in e847f79f1a437b15cee7688dbd65418d1661f7a3..0e87dbfc20927e824a004468783155641970c354.

I'd love some feedback on:
* Whether the runtime validation is, well, valid
 * More test cases?
 * Am I manipulating AST values in the correct manner? (I feel like I'm doing a lot just to validate types and such)
* How to best incorporate this in the docs (update `docs\content\policy_reference.md` directly?)

Reply:
Ahh, I see. I think that for this built-in it makes sense to add a small note in the built in description about the available fields since the number is small. So just this part for now:


<img width="722" alt="Screenshot 2024-03-28 at 09 30 46" src="https://github.com/open-policy-agent/opa/assets/1774239/8f39ba8f-deaf-484e-8d9f-e239286c9e1d">

Do you agree @johanfylling ?

Reply:
Regarding static properties: 
I think I got that wrong the first time around.
If we look at [this example with json.patch()](https://play.openpolicyagent.org/p/VDlsyKaSKc): 

```rego
package play

p1 := json.patch({"a": {"foo": 1}}, [{"op": "add", "path": "/a/bar", "value": 2}])

p2 := json.patch({"a": {"foo": 1}}, [{"path": "/a/bar", "value": 2}])

p3 := json.patch({"a": {"foo": 1}}, [{"op": 2, "path": "/a/bar", "value": 2}])
```

`p2` is omitting the static `op` property without complaints.
Whereas `p3` is assigning it the wrong type (`number` instead of `string`).
The benefit here is compile-time type checking.

Regarding the default behavior of the built-in: 
since we're calling this `json.marshal_with_options` now, user expectations might be that it'll behave exactly like `json.marshal` if no options are provided. This would however have the unfortunate side effect of requiring an additional `opt` property (like a boolean `pretty` toggle) to turn on indentation if `indent` or `prefix` isn't assigned. Seen in isolation from `json.marshal`, I think it doesn't matter much if the default is pretty printing or not, as long as there is an option to enable/disable. Thoughts?

I swear, my intention is not to bikeshed this PR 😅.

Reply:
So, I tried implementing that as follows:
```go
var JSONMarshalWithOptions = &Builtin{
	Name: "json.marshal_with_options",
	Description: "Serializes the input term to multiline JSON, with specified formatting options (indent/prefix). " +
		"One or more copies of the indent string are emitted before each child JSON element, depending on indent level.",
	Decl: types.NewFunction(
		types.Args(
			types.Named("x", types.A).Description("the term to serialize"),
			types.Named("opts", types.NewAny(
				types.NewNull(),
				types.NewObject(
					[]*types.StaticProperty{
						types.NewStaticProperty("indent", types.NewAny(types.S, types.NewNull())),
						types.NewStaticProperty("prefix", types.NewAny(types.S, types.NewNull())),
					},
					nil,
				),
			)).Description("encoding options - accepts keys `prefix` (string to prefix lines with, default empty string) and `indent` (string to indent with, default `\\t`)"),
		),
		types.Named("y", types.S).Description("the JSON string representation of `x`, with configured prefix/indent string(s) as appropriate"),
	),
	Categories: encoding,
}
```

This caused test `jsonbuiltins/marshal_with_options-missing-key` (which defines `prefix` but not `indent`) to fail. What perplexes me is that allowing dynamic properties (i.e. replace `nil` with `types.NewDynamicProperty(types.S, types.A)` fixed that test, but that also means `jsonbuiltins/marshal_with_options-invalid-key` resorts to runtime checks.

FWICT, it seems that the static property type-checking still functions, but there's some weird overlap between static/dynamic properties. I see two potential problems here:
* If dynamic properties are intentionally allowed, you can't "require" static properties to be present (though I guess the counterpoint is that those should just be params)
* If dynamic properties aren't intentionally allowed (and you're just doing it to make statics optional), you can't assert that objects only contain a set of "good" keyvalues and have to resort to runtime checks on at least the keys themselves (if not the values)

I'm not sure if this is a bug or not. The behavior was definitely a little confusing. I converted to static properties (with dynamics still enabled) in 54ee49326dd720e6d2cdc574e2c2446732577c6e.

Reply:
> Regarding the default behavior of the built-in: since we're calling this `json.marshal_with_options` now, user expectations might be that it'll behave exactly like `json.marshal` if no options are provided. This would however have the unfortunate side effect of requiring an additional `opt` property (like a boolean `pretty` toggle) to turn on indentation if `indent` or `prefix` isn't assigned. Seen in isolation from `json.marshal`, I think it doesn't matter much if the default is pretty printing or not, as long as there is an option to enable/disable. Thoughts?

@johanfylling I see the potential for confusion. The other option that comes to mind (without having to add more keys to the options structure) is calling the builtin `marshal_formatted` or `marshal_pretty` - this might sidestep the ambiguity entirely.

Reply:
@johanfylling I was thinking about the idea of a third `pretty` option - another potential hiccup that came to mind was conflicting settings between the three options. If someone calls `json.marshal_with_options(x, {"pretty": false, "indent": "  "}`, I'm not sure what we'd want to do there. 

We could throw an error since the parameters don't make sense with one-another, but in-general it feels like it'd get cumbersome (especially if more options were added later).

HDYF about renaming the builtin so it's clear this is strictly for formatted/pretty-printed JSON?

Reply:
I don't think there should be much ambiguity if `pretty` is set to `false` at the same time as other parameters are given a value, as that is pretty explicitly turning off pretty formatting. I could imagine a scenario where you do something like `json.marshal_with_options(x, {"pretty": input.pretty_print, "indent": "  "})` where the formatting particulars are constant, but pretty formatting is turned on based on evaluation circumstances.

What I could see as more ambiguous is if `pretty` is _not_ declared, but e.g. `indent` is. It'd make sense to let declared `indent` and/or `prefix` parameters imply `pretty: true`, though, as that would also provide the option to make the call a bit less verbose in some cases.

Even though it's pretty unlikely that json marshaling will see future configurable options (maybe there is a slim chance that someone might want e.g. different string escaping mechanisms added), I think the `*_with_options({...})` is a pretty good pattern that we might want to establish for other, future built-ins, and what you're doing here is an awesome start.

Reply:
@johanfylling Okay, I've added a `pretty` option in e09f43a69d0ef5fd9b7ec5f0b35477f02b71b1e2.

> I think the `*_with_options({...})` is a pretty good pattern

If this is the case, it would significantly aide implementation of these builtins if we had a helper to convert `ast.Object` to a `map[string]any` (or even a `map[any]any`?), optionally accepting an array of "approved" keys, and returning `(map[...]any, error)` with some helpful text.

Currently, roughly 1/4 of `encoding.go` is this specific function - most of it is just trying to massage values out of the AST. If more `*_with_options` functions are expected, I expect this boilerplate would make more complex builtins very hard to read.

Also, I'm still interested in your thoughts on the static/dynamic behavior I described in https://github.com/open-policy-agent/opa/pull/6636#discussion_r1543741952 - this doesn't directly block implementation on this builtin, but having static properties treated like dynamic ones seems like a pitfall other `*_with_options` implementers would fall into.

Reply:
Yes, adding some sort of helper to assert the content of an options object would definitely be helpful for future built-ins employing this pattern 👍 . I'll create a separate ticket for tracking something like this.

Regarding dynamic/static typing of objects: I agree, it's not intuitive that adding a dynamic component to the object would make static properties optional, and leaving it out makes them required. I don't know if there is a current means to get exactly the behavior we want here (perhaps this can be improved upon in the type checker). Your current solution seems to be the appropriate one; a dynamic property paired with run-time checks.

I suppose it would also be possible to define the `opts` type to be an `any` containing objects with only static properties for every possible permutation: `{"pretty": B, "index": N, "indent": N}`, `{"pretty": B}`, `{"pretty": B, "index": N}`, etc. But this is extremely unwieldy with just these three options, and wouldn't scale at all.

Reply:
Sounds good. Do you think implementation of an options-object-assertion helper needs to happen before this PR can be merged, or can that be handled in a later release?

(Just want to make sure I've addressed all of the feedback in this PR before it gets a ✅)

Reply:
I see no reason to delay this PR because of that.

Reply:
I should also add that if we adopt the `*_with_options` pattern more frequently, we should figure out how to better present the `opts` objects in the docs.

With `marshal_with_options`, the formal type of `opts` is so long that text appears scrunched. (It was even worse when the description was longer 😅)
![image](https://github.com/open-policy-agent/opa/assets/72675818/6b89e108-9c86-48e5-9ec3-a08a0cc8e5bf)

With `x509.parse_and_verify_certificates_with_options`, the type declaration is shorter but the lack of flexibility in AST description tags means the type description is difficult to read without formatting:

![image](https://github.com/open-policy-agent/opa/assets/72675818/7f24b430-147b-49e9-b118-e3752e801587)

Reply:
> I see no reason to delay this PR because of that.

@johanfylling Sorry, missed the earlier message. What's the next steps for getting this PR approved? I'm not sure who else needs to review, or what the signoff process looks like.

Reply:
I wonder if we need the `types.NewAny` wrappers on the parameters, or if it's enough to just use `types.S` and `types.B` where applicable 🤔.

Reply:
Any green review from a maintainer is enough for the PR to be merged. If no one beats me to it, I'll try to go through it all today or tomorrow.